### PR TITLE
subChannel should be able to set peer preference

### PR DESCRIPTION
**Pull request by ShanniLi**
r @jcorbin @Raynos @kriskowal @anson627

### CODE REVIEW

File: node/peer.js [hunk 1]
```diff
@@ -65,14 +63,8 @@ function TChannelPeer(channel, hostPort, options) {
...
-    self.preferConnectionDirection = self.options.preferConnectionDirection;
-    if (self.preferConnectionDirection === 'out') {
-        self.setScoreStrategy(PreferOutgoing);
-    } else if (self.preferConnectionDirection === 'in') {
-        self.setScoreStrategy(PreferIncoming);
-    } else {
-        self.setScoreStrategy(NoPreference);
-    }
+    self.heapElementCollection = {};
+    self.handlers = {};
```
Comment (anson627) [inline node/peer.js]: it's score_strategy instead of hander, right?

File: node/peer_heap.js [hunk 2]
```diff
@@ -24,13 +24,15 @@ module.exports = PeerHeap;
...
-function PeerHeap() {
+function PeerHeap(direction) {
     var self = this;
 
     self.array = [];
     // TODO: worth it to keep a tail free list like TimeHeap?
     // self.end = 0;
     self._stack = [];
+
+    self.direction = direction;
```
Comment (anson627) [inline node/peer_heap.js]: is peer heap allowed to have peers with different directions?

File: node/sub_peers.js [hunk 3]
```diff
@@ -63,11 +64,17 @@ TChannelSubPeers.prototype.add = function add(hostPort, options) {
 
     peer = topChannel.peers.add(hostPort, options);
 
+    peer.addScoreStrategy(self.preferConnectionDirection);
     self._map[hostPort] = peer;
     self._keys.push(hostPort);
 
     var el = self._heap.add(peer);
-    peer.heapElements.push(el);
+    var heapElements = peer.heapElementCollection[self.preferConnectionDirection];
```
Comment (anson627) [inline node/sub_peers.js]: how about having a method called getOrCreateHeapElements()?

File: node/sub_peers.js [hunk 4]
```diff
@@ -91,11 +98,12 @@ TChannelSubPeers.prototype._delete = function _del(peer) {
     delete self._map[peer.hostPort];
     popout(self._keys, index);
 
-    for (var i = 0; i < peer.heapElements.length; i++) {
-        var el = peer.heapElements[i];
+    var heapElements = peer.heapElementCollection[self.preferConnectionDirection] || [];
```
Comment (anson627) [inline node/sub_peers.js]: so this can be replaced with getOrCreateHeapElements.

File: node/peer.js (hunk 1 above)
Comment (ShanniLi) [inline node/peer.js]: I used the older name "handler". In this case, "handler" ==> "handlers"

File: node/peer_heap.js (hunk 2 above)
Comment (ShanniLi) [inline node/peer_heap.js]: no, otherwise, the heap won't work.

File: node/sub_peers.js (hunk 3 above)
Comment (ShanniLi) [inline node/sub_peers.js]: Good point!

**Comment by anson627** [comment]
:+1:

File: node/peer.js [hunk 5]
```diff
@@ -65,14 +63,8 @@ function TChannelPeer(channel, hostPort, options) {
...
 
-    self.preferConnectionDirection = self.options.preferConnectionDirection;
-    if (self.preferConnectionDirection === 'out') {
-        self.setScoreStrategy(PreferOutgoing);
-    } else if (self.preferConnectionDirection === 'in') {
-        self.setScoreStrategy(PreferIncoming);
-    } else {
-        self.setScoreStrategy(NoPreference);
-    }
+    self.heapElementCollection = {};
```
Comment (kriskowal) [inline node/peer.js]: heapElementsByStrategy

File: node/peer.js [hunk 6]
```diff
@@ -95,23 +87,36 @@ function TChannelPeer(channel, hostPort, options) {
...
+TChannelPeer.prototype.addScoreStrategy = function addScoreStrategy(direction) {
     var self = this;
 
-    self.handler = new ScoreStrategy(self);
+    if (direction === 'out') {
+        self.handlers.out = new PreferOutgoing(self);
+    } else if (direction === 'in') {
+        self.handlers.in = new PreferIncoming(self);
+    } else {
+        self.handlers.default = new NoPreference(self);
```
Comment (kriskowal) [inline node/peer.js]: neutral

File: node/peer.js [hunk 7]
```diff
@@ -95,23 +87,36 @@ function TChannelPeer(channel, hostPort, options) {
...
 
+
+
 TChannelPeer.prototype.invalidateScore = function invalidateScore() {
     var self = this;
 
-    if (!self.heapElements.length) {
-        return;
-    }
+    var directions = Object.keys(self.handlers);
```
Comment (kriskowal) [inline node/peer.js]: static. add to prototype.

File: node/peer.js [hunk 8]
```diff
@@ -215,16 +220,16 @@ TChannelPeer.prototype.countConnections = function countConnections(direction) {
...
 
 // ensures that a connection exists
-TChannelPeer.prototype.connect = function connect(outOnly) {
+TChannelPeer.prototype.connect = function connect(preferConnectionDirection) {
     var self = this;
     var conn = null;
-    if (self.preferConnectionDirection === 'in' && !outOnly) {
+    if (preferConnectionDirection === 'in') {
         conn = self.getIdentifiedInConnection();
     } else {
```
Comment (kriskowal) [inline node/peer.js]: add comment // 'out' or 'neutral'

File: node/peer.js [hunk 9]
```diff
@@ -443,15 +448,35 @@ TChannelPeer.prototype.countPending = function countPending(direction) {
 // Called on connection change event
 TChannelPeer.prototype._maybeInvalidateScore = function _maybeInvalidateScore() {
     var self = this;
+    var should = false;
 
-    if (self.handler.getTier() !== self.handler.lastTier) {
+    var directions = Object.keys(self.handlers);
```
Comment (kriskowal) [inline node/peer.js]: static

**Comment by kriskowal** [comment]
review feedback is advisory. please ensure that @jcorbin weighs in.

**Comment by Raynos** [comment]
Discussed with @jcorbin out of band.

Decision was to not touch the heap and make each peer have only one preference.

Instead we want to update ServiceProxy and HyperbahnClient to set the correct preference at "peer creation time". We may need some extra args to `makeSubChannel` to pass preference through to the createPeer() call.

**Comment by ShanniLi** [comment]
moved to #1348

**Pull request closed without merging**